command/agent_test.go

```diff
@@ -3097,6 +3100,323 @@ vault {
 	wg.Wait()
 }
 
+func setupTestApp(t *testing.T) string {
+	t.Helper()
+
+	// we must build a test-app binary since 'go run' does not propagate signals correctly
+	goBinary, err := exec.LookPath("go")
+	if err != nil {
+		t.Fatalf("could not find go binary on path: %s", err)
+	}
+
+	testAppBinary := filepath.Join(os.TempDir(), "test-app")
+
+	if err := exec.Command(goBinary, "build", "-o", testAppBinary, "./agent/exec/test-app").Run(); err != nil {
+		t.Fatalf("could not build the test application: %s", err)
+	}
+	return testAppBinary
+}
+
+// TestAgent_Exec_Restarts tests that vault agent restarts the app
+// on changes to relevant vault secrets
+func TestAgent_Exec_Restarts(t *testing.T) {
+	testAppBin := setupTestApp(t)
+	defer os.Remove(testAppBin)
+
+	vaultClient, cleanup := testVaultServer(t)
+	defer cleanup()
+	t.Setenv(api.EnvVaultAddress, vaultClient.Address())
+
+	tokenFile := populateTempFile(t, "tokenfile.txt", vaultClient.Token())
+	defer os.Remove(tokenFile.Name())
+
+	// token file needs to have 600 permissions
+	if err := os.Chmod(tokenFile.Name(), 0o600); err != nil {
+		t.Fatalf("unable to change permissions of token file: %s", err)
+	}
+
+	const port = 34001
+	config := fmt.Sprintf(`
+template_config {
+  static_secret_render_interval = "2s"
+}
+
+vault {
+  address = "%s"
+  tls_skip_verify = true
+}
+
+auto_auth {
+	method {
+		type = "token_file"
+    	config {
+      		token_file_path = "%s"
+		}
+	}
+}
+env_template "MY_DATABASE_USER" {
+  contents = "{{ with secret \"kv-v2/data/my-db\" }}{{ .Data.data.user }}{{ end }}"
+}
+env_template "MY_DATABASE_PASSWORD" {
+  contents = "{{ with secret \"kv-v2/data/my-db\" }}{{ .Data.data.password }}{{ end }}"
+}
+
+exec {
+	command = ["%s", "-port", "%d"]
+}`, vaultClient.Address(), tokenFile.Name(), testAppBin, port)
+	logger := logging.NewVaultLogger(hclog.Trace)
+	configFile := makeTempFile(t, "config.hcl", config)
+
+	// enable kv-v2 backend
+	if err := vaultClient.Sys().Mount("kv-v2/", &api.MountInput{
+		Type: "kv-v2",
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	_, err := vaultClient.KVv2("kv-v2/").Put(context.Background(), "my-db", map[string]interface{}{
+		"user":     "dbuser",
+		"password": "password1",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, agentCmd := testAgentCommand(t, logger)
+	agentCmd.client = vaultClient
+	agentCmd.startedCh = make(chan struct{})
+
+	// the command blocks, so run in go-routine
+	agentExitCodeCh := make(chan int, 1)
+	go func(exitCodeCh chan<- int) {
+		args := []string{"-config", configFile}
+		exitCodeCh <- agentCmd.Run(args)
+	}(agentExitCodeCh)
+
+	// wait until agent starts
+	select {
+	case <-agentCmd.startedCh:
+	case <-time.After(10 * time.Second):
+		t.Fatal("agent did not start within 10 seconds")
+	}
+
+	// agent started, give some time to populate env vars from vault
+	time.Sleep(10 * time.Second)
```

Comment:
10s seems excessive. Maybe try retryablehttp to poll it like we did [in exec_test.go](https://github.com/hashicorp/vault/blob/main/command/agent/exec/exec_test.go#L300)?

Reply:
its a little different... just waiting for consul-template to render. will change to something smaller tho